src/ILLink.RoslynAnalyzer/DataFlow/LocalDataFlowVisitor.cs

```diff
@@ -180,6 +199,12 @@ public static IMethodSymbol GetPropertyMethod (IPropertyReferenceOperation opera
 
 		public override TValue VisitPropertyReference (IPropertyReferenceOperation operation, LocalDataFlowState<TValue, TValueLattice> state)
 		{
+			if (operation.Parent is IFlowCaptureOperation) {
```

Comment:
Couldn't we just avoid visiting the child operation in the parent?

Although this does raise the question of what the behavior should be for nested property references, i.e.

```C#
Prop1.Prop2 = (a ?? b)
```

Reply:
I removed this check and it should now be handled in the parent. For nested properties, I'm not sure if this is exactly what you are referring to - but I think the semantics we ideally want are that:
1. Visit the LHS to produce a reference to the property along with the object instance. (It should only visit Prop1 once in the whole process)
2. Visit the RHS to get the value of `a ?? b`
3. Assign the value through the saved PropertyReference.

Instead, what we currently do is to avoid visiting the LHS in `VisitFlowCapture` (instead we capture the "unevaluated" PropertyReferenceOperation), and visit it after the RHS. That's for the simple case where the property is only captured as an LValue. If it's captured as an LValue and RValue, we will actually visit it both during the capture and during the assignment.

It's not ideal, but I couldn't think of any cases where this would impact dataflow warnings (but there certainly might be ones I'm not thinking of). I would like to fix this more properly by tracking more information in the flow capture, such as the object instance, but I'm not sure yet what additional information we might need to track in other cases. I think this approach is good enough for now, and we can fix it if we find problems, or if the work overlaps with the solution to https://github.com/dotnet/linker/issues/2158 - how does that sound?

Reply:
I left some comments about it in the code too - PTAL.